tablets: support repairing a single table in a co-location group

mlitvk opened the pull request:
With the recent introduction of co-located tables, A tablet map can be shared between multiple tables.This created a limitation in the way we can repair tablets, due to the way the repair info is stored in the tablet map:

* The repair_time for each tablet is stored in the tablet map, which means it is now a shared repair_time of all co-located tablets that share this tablet map.
* The repair task info is also stored in the shared tablet map, so it cannot represent a task to repair the tablet of a specific table within a co-location group.

The way to address this was to consider tablet repair as an operation that works on group of co-located tablets and not on tablets of a specific table. We can only request to repair an entire co-located tablet group, and they will all have a shared repair time.

This is not ideal, because if we have a base table with many co-located views, for example, and we want to repair only the base table or a single view - we can't, we must run repair on all. It also complicates the API for users, because they need to know whether a table is co-located or not.

To fix this, we bring back support to request repair for the tablets of a specific table, even if it's co-located with other tables. We change `repair_time` to be a map, so we can store an independent repair time for each tablet in the co-location group, and we add a `tables_filter` value to the repair request, so a specific table can be requested for repair.

Fixes scylladb/scylladb#24816

mykaul commented:
IIRC, @patjed41 was chasing slow flushes. How slow and on which server did the test run?

mlitvk commented:
on spider 11

[test_move_tablet[base].debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/18142/testReport/junit/cluster/test_tablets_colocation/debug___test_py___debug___test_move_tablet_base__debug_1): almost 2 minutes
```
INFO  2025-07-06 05:20:31,752 [shard 0:main] init - replaying commit log - flushing memtables
INFO  2025-07-06 05:22:18,664 [shard 0:main] init - replaying commit log - removing old commitlog segments
```

[test_move_tablet[base].debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/18142/testReport/junit/cluster/test_tablets_colocation/debug___test_py___debug___test_move_tablet_base__debug_3): almost 3 minutes
```
INFO  2025-07-06 05:17:29,954 [shard 0:main] init - replaying schema commit log - flushing memtables
INFO  2025-07-06 05:20:16,821 [shard 0:main] init - replaying schema commit log - removing old commitlog segments
```

[test_move_tablet[child].debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/18142/testReport/junit/cluster/test_tablets_colocation/debug___test_py___debug___test_move_tablet_child__debug_1): the last line is
```
INFO  2025-07-06 05:32:51,503 [shard 0:main] init - replaying schema commit log - flushing memtables
```

[test_move_tablet[child].debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/18142/testReport/junit/cluster/test_tablets_colocation/debug___test_py___debug___test_move_tablet_child__debug_3): the last line is
```
INFO  2025-07-06 05:32:09,681 [shard 0:main] init - replaying schema commit log - flushing memtables
```

patjed41 commented:
> IIRC, @patjed41 was chasing slow flushes. How slow and on which server did the test run?

Me and @bitpathfinder were chasing slow flushes on spider9 and spider11.

#12028 is the issue for all tests failing due to slow flushes during shutdown. However, here the test failed due to slow flushes during startup. It's the first time I'm seeing such a failure. The root cause it probably the same, but I'll open a new issue just in case.

patjed41 commented:
> The root cause it probably the same, but I'll open a new issue just in case.

I opened #24867

mykaul commented:
@mlitvk - any updates? The test is still failing us on CI.

mlitvk commented:
> @mlitvk - any updates? The test is still failing us on CI.

I will discuss this with @piodul

mlitvk commented:
Updated to allow repairing the tablets of co-located tables independently
* Changed the `repair_time` in the tablet map to a `repair_times` maps, to allow repairing a single table in a co-location group and have independent repair time for each.
* Added `tables_filter` to the repair task info to allow specifying a specific table to repair in the user repair request

This fixes the issue by allowing repair requests for co-located tables without breaking the API and without amplification by unintended repairs

asias commented on docs/dev/system_keyspace.md:
```diff
@@ -279,7 +292,9 @@ Only tables which use tablet-based replication strategy have an entry here.
-`repair_time` is the last time the tablet has been repaired.
+`repair_time` is the last time the tablet has been repaired. It is deprecated and replaced by `repair_times`.
+
+`repair_times` is a map that maps a table to the last time the table's tablet has been repaired. Since the tablet map can be shared by multiple co-located tables, each table that shares the tablet map can have a different repair time.
```
At a higher layer, can we make the tablet_map into too parts, 1) that are shared between tables 2) that that are not shared between tables. This way we can avoid hacking individual member field that has to be per tablet. 

Per tablet is a huge advantage that tablet gives us, let's not break the fundamentals for a sub feature of tablets.

mlitvk replied:
I'm looking into this, thinking how to store the repair information per table, thanks
originally we wanted the tablet map to be entirely shared to avoid any kind of duplication or inconsistencies that could happen, but I agree it's not great for the repair information

tgrabiec replied:
Why not store non-shared parts like repair time in system.tablets under the partition key of the co-located table? The "base_table" static column still indicates that the shared parts are shared, not shared parts are read from the partition.

mlitvk commented on docs/dev/system_keyspace.md:
```diff
@@ -256,6 +258,17 @@ CREATE TYPE system.tablet_task_info (
+    sched_time timestamp,
+    repair_hosts_filter text,
+    repair_dcs_filter text,
+    tables_filter text
```
the idea was that since the tablet map is now shared in general between multiple tables, the tablet_task_info doesn't belong to a tablet of a specific table and we don't know which table was requested for repair.

bhalevy commented:
What's the use case of repairing a co-located non-base table?

bhalevy commented:
> > What's the use case of repairing a co-located non-base table?
> 
> repairing a co-located MV table, for example. For the other co-located tables - CDC and paxos table, I'm not sure if there's case for repair, but I think there's still need for MV repair if I understand correctly.

See also https://github.com/scylladb/scylladb/issues/19727
I don't see much point in repairing the view before repairing the base table and waiting for view update generation to complete (then repair the view)

mykaul commented:
> @nyh - please take a look.

re-ping @nyh - please add your thoughts on this.

bhalevy commented:
> @nyh - please take a look.

@nyh ping.  We need your input as well.

mlitvk commented:
> > > What's the use case of repairing a co-located non-base table?
> > 
> > 
> > repairing a co-located MV table, for example. For the other co-located tables - CDC and paxos table, I'm not sure if there's case for repair, but I think there's still need for MV repair if I understand correctly.
> 
> See also #19727 I don't see much point in repairing the view before repairing the base table and waiting for view update generation to complete (then repair the view)

The goal of the PR is to allow repairing each table independently, even if it's co-located, as was before.
Currently for co-located tables you can only request repair for the entire group, not for specific tables, due to the limitation in the way the repair fields are shared.
This will also allow to repair a base table and its co-located views independently, or in a specific order, etc. It will become an orthogonal issue.

bhalevy commented:
> > > > > > What's the use case of repairing a co-located non-base table?
> > > > > 
> > > > > 
> > > > > repairing a co-located MV table, for example. For the other co-located tables - CDC and paxos table, I'm not sure if there's case for repair, but I think there's still need for MV repair if I understand correctly.
> > > > 
> > > > 
> > > > See also #19727 I don't see much point in repairing the view before repairing the base table and waiting for view update generation to complete (then repair the view)
> > > 
> > > 
> > > The goal of the PR is to allow repairing each table independently, even if it's co-located, as was before. Currently for co-located tables you can only request repair for the entire group, not for specific tables, due to the limitation in the way the repair fields are shared. This will also allow to repair a base table and its co-located views independently, or in a specific order, etc. It will become an orthogonal issue.
> > 
> > 
> > It is a step backwards IMO. First, let's first adjust the test to the existing behavior to stabilize CI, then we can discuss further changes to death :)
> 
> I can "fix" the test very easily by making it request repair for the base table (which will repair the base table and the co-located view) instead of requesting repair for the view table. The problem is I'm not this is the behavior we want to test and if it makes sense for the test to do this. or we can just skip the test until the issue is closed.

What does git history for the change in scylla reveal?
